test/Kestrel.FunctionalTests/ResponseTests.cs

```diff
@@ -294,6 +300,98 @@ public Task ResponseStatusCodeSetBeforeHttpContextDisposedRequestMalformedReadIg
+                    {
+                        context.Response.OnCompleted(_ => throw new Exception(), null);
+                        await context.Response.WriteAsync("hello, world");
+                    });
+                });
+
+            using (var host = hostBuilder.Build())
+            {
+                host.Start();
+
+                using (var client = new HttpClient())
+                {
+                    var response = await client.GetAsync($"http://127.0.0.1:{host.GetPort()}/");
+                    Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+                }
+            }
+        }
+
+        [Fact]
+        public async Task OnCompletedShouldNotBlockAResponse()
+        {
+            var delay = Task.Delay(TestConstants.DefaultTimeout);
+            var hostBuilder = TransportSelector.GetWebHostBuilder()
+                .UseKestrel()
+                .UseUrls("http://127.0.0.1:0/")
+                .Configure(app =>
+                {
+                    app.Run(async context =>
+                    {
+                        context.Response.OnCompleted(async () =>
+                        {
+                            await delay;
+                        });
+                        await context.Response.WriteAsync("hello, world");
+                    });
+                });
+
+            using (var host = hostBuilder.Build())
+            {
+                host.Start();
+
+                using (var client = new HttpClient())
+                {
+                    var response = await client.GetAsync($"http://127.0.0.1:{host.GetPort()}/");
+                    Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+                    Assert.False(delay.IsCompleted);
+                }
+            }
+        }
+
+        [Fact]
+        public async Task InvalidChunkedEncodingInRequestShouldNotBlockOnCompleted()
+        {
+            var onCompletedCalled = false;
+            using (var server = new TestServer(httpContext =>
+            {
+                httpContext.Response.OnCompleted(_ =>
+                {
+                    onCompletedCalled = true;
+                    return Task.CompletedTask;
+                }, null);
+                return Task.CompletedTask;
+            }))
+            {
+                using (var connection = server.CreateConnection())
+                {
+                    await connection.Send(
+                        "GET / HTTP/1.1",
+                        "Host:",
+                        "Transfer-Encoding: chunked",
+                        "",
+                        "gg");
+                    await connection.ReceiveForcedEnd(
+                        "HTTP/1.1 200 OK",
+                        $"Date: {server.Context.DateHeaderValue}",
+                        "Content-Length: 0",
+                        "",
+                        "");
+                }
+            }
+            Assert.True(onCompletedCalled);
```

Comment:
I think you want to use a TaskCompletionSource here like you did for the OnCompleteCalledEvenWhenOnStartingNotCalled test since the response can be fully written prior to OnCompleted being called.